### ISSUE TITLE
Delay on adding/removing printers for qz.printers().find()

### ISSUE THREAD

**Issue by majoncas** [qzind/tray#393]
I figured out the problem in our application then I tried to reproduce the problem on your demo page on your website so I can proof that this is not an implementation problem.

On my MAC, I add a new printer, then I click on "Disconnect" and after I click on "Connect" to force a new QZ connection. I click on "Find all printers" button and my newly added printer is not visible. I need to do the same thing a lot of time before being able to see it in the list returned by QZ. I can reproduce the same scenario when I delete the printer it takes a lot of tries before my deleted printer becomes unlisted. Is there any caching on QZ side which prevent displaying up to date information regarding available printers list ?

Thank you for your hep.

**Comment by tresf** [comment]
> Is there any caching on QZ side which prevent displaying up to date information regarding available printers list ?

No, we get this information from CUPS which -- especially on macOS -- is not realtime.  I recommend you have the printers configured before starting QZ Tray and restarting between changes.

**Comment by majoncas** [comment]
I just realized that as a workaround, closing the QZ Tray application and restart it would fix the issue but I would not expect as a user to do such operation in order to make this work. The API should send me accurate informations.

**Comment by tresf** [comment]
> The API should send me accurate informations.

We use standard Java APIs, the issue is not with our product.

**Comment by majoncas** [comment]
It's ok then. I just wanted to check with you guy if this was really "as designed" . Thank you for your confirmation

**Comment by CindyLulu** [comment]
@majoncas , I have the same issue too, add below code into the first line of PrintServiceMatcher getPrintServices function can help resolve this issue:
AppContext.getAppContext().put(PrintServiceLookup.class.getDeclaredClasses()[0], null);

**Comment by tresf** [comment]
Reopening to see if we can improve this logic.

**Comment by klabarge** [comment]
@CindyLulu, thanks for providing this!

I've added
`AppContext.getAppContext().put(PrintServiceLookup.class.getDeclaredClasses()[0], null);` to the below line and it indeed speeds up the printer listing in CUPS. 

When a printer is added or removed, the first call to `qz.printers.find` reflects the changes (without any refreshing) 👍 

https://github.com/qzind/tray/blob/911c525e5ad9f0042eaa63fe1158ef4104a1a013/src/qz/printer/PrintServiceMatcher.java#L28

I'll commit this into our code, and it will make into the 2.0.10 release

**Comment by tresf** [comment]
Turns out the solution in 023843e uses a proprietary internal API and may be removed from future Java versions.  We need to file an upstream bug with AdoptOpenJDK11 to patch this in the JDK so that we don't get bit by it down the road.

Developers: If IntelliJ complains `package sun.awt does not exist`, see the workaround here: https://stackoverflow.com/a/56388149/3196753

**Comment by tresf** [comment]
If Java removed the `sun.awt` API, a potential solution is to use Java reflection to try to set this property... 

```java
/**
 * Updates printer listing in CUPS immediately
 */
public static void forcePrintersRefresh() {
    try {
        //Update printer list in CUPS immediately
        Class appContextClass = Class.forName("sun.awt.AppContext");
        Method getAppContextMethod = appContextClass.getMethod("getAppContext");
        getAppContextMethod.setAccessible(true);
        Object appContextInstance = getAppContextMethod.invoke(appContextClass);
        Method putMethod = appContextInstance.getClass().getMethod("put", Object.class, Object.class);
        putMethod.invoke(appContextInstance, PrintServiceLookup.class.getDeclaredClasses()[0], null);
        AppContext.getAppContext().put(PrintServiceLookup.class.getDeclaredClasses()[0], null);
    } catch (Throwable t) {
        // Catches JDK9+ "InaccessibleObjectException"
    }
}
```

The downside to this approach is that the JDK9+ warns:

```diff
- WARNING: An illegal reflective access operation has occurred
- WARNING: Illegal reflective access to method sun.awt.AppContext.getAppContext()
- WARNING: Please consider reporting this to the maintainers of test.Main
- WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
- WARNING: All illegal access operations will be denied in a future release
```

Testing the code with `--illegal-access=deny` actually shows no warnings because the `InaccessibleObjectException` is swallowed by the `Throwable` despite it being a JDK9+ class.

For now, we can keep this proprietary call, but I'll ask @klabarge to file a bug report upstream so that we can get it fixed properly and (eventually) remove the workaround.

**Comment by CindyLulu** [comment]
Sorry @tresf , after few days print trying, I find out that print will become very slow after one or two days until restarting the client. I checked the client log, it took more than 2 minutes to excuting until getting the remote endpoint unavailable error.
`2019-06-28 09:49:25,331 [DEBUG] Executing: [lpstat, -l, -p, DYMO_LabelWriter_450_Twin_Turbo]
2019-06-28 09:49:25,392 [DEBUG] Executing: [grep, *PCFileName:, /private/etc/cups/ppd/DYMO_LabelWriter_450_Twin_Turbo.ppd]
2019-06-28 09:49:25,740 [DEBUG] Found Resolution: 30000x30000 dphi
2019-06-28 09:49:25,748 [DEBUG] Executing: [lpstat, -l, -p, Zebra_ZTCTLP2844-Z-200dpi-ZPL]
2019-06-28 09:49:25,840 [DEBUG] Executing: [lpoptions, -p, Rollo_Printer, -l]
2019-06-28 09:49:26,149 [DEBUG] Parsed default density from CUPS Rollo_Printer: 203dpi
2019-06-28 09:49:26,150 [DEBUG] Executing: [lpoptions, -p, Zebra_ZTCTLP2844-Z-200dpi-ZPL, -l]
2019-06-28 09:49:26,456 [DEBUG] Executing: [lpoptions, -p, DYMO_LabelWriter_450_Twin_Turbo, -l]
2019-06-28 09:49:26,751 [DEBUG] Parsed default density from CUPS DYMO_LabelWriter_450_Twin_Turbo: 300dpi
2019-06-28 09:49:26,751 [DEBUG] Executing: [lpoptions, -p, rollo_PDFwriter, -l]
2019-06-28 09:49:27,048 [DEBUG] Executing: [lpoptions, -p, Zebra_Technologies_ZTC_TLP2844_Z_200dpi, -l]
2019-06-28 09:49:27,359 [DEBUG] Parsed default density from CUPS Zebra_Technologies_ZTC_TLP2844_Z_200dpi: 203dpi
2019-06-28 09:49:27,359 [ERROR] Problem processing message
org.eclipse.jetty.websocket.api.WebSocketException: RemoteEndpoint unavailable, outgoing connection not open
	at org.eclipse.jetty.websocket.common.WebSocketSession.getRemote(WebSocketSession.java:251)
	at qz.ws.PrintSocketClient.send(Unknown Source)
	at qz.ws.PrintSocketClient.sendResult(Unknown Source)
	at qz.ws.PrintSocketClient.processMessage(Unknown Source)
	at qz.ws.PrintSocketClient.onMessage(Unknown Source)
	at jdk.internal.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
	at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.base/java.lang.Thread.run(Thread.java:835)` 

This issue is resolved by removing the below code change: AppContext.getAppContext().put(PrintServiceLookup.class.getDeclaredClasses()[0], null); 
I doubt this code will bring performance issue , and will this issue (https://github.com/qzind/tray/issues/454) also relate to this change?

**Comment by tresf** [comment]
> This issue is resolved by removing the below code change: `AppContext.getAppContext().put(PrintServiceLookup.class.getDeclaredClasses()[0], null);`
I doubt this code will bring performance issue

I don't understand the question.  You say it fixes performance but then state that you doubt it will fix performance?  We test on macOS too and we don't see a huge performance problem.  I'm surprised that only a few printers are causing this problem.  Does it still happen if you reset the printing subsystem (warning, this will remove all printers)?

>  will this issue (#454) also relate to this change?

No, it's not related.  #454 is related to resource allocation/freeing for large image rasterization.

**Comment by tresf** [comment]
> after few days print trying, I find out that print will become very slow in Mac after one or two days until restarting the client. I checked the client log, it took more than 2 minutes to excuting until getting the remote endpoint unavailable error.

@CindyLulu sorry for not taking your comments seriously.  You're right.  We've identified a major regression as a result of this patch and we're reopening. 

Commit https://github.com/qzind/tray/commit/023843e56da9ab8307a5334601e7ae3c5bf32ab7 will need to be reverted.  If we cannot find a better solution, we'll have to file a bug report upstream.

Reverting to 2.0.9 is a temporary solution.  We'll file a dedicated bug report for the regression.

**Comment by tresf** [comment]
Upstream bug report filed: https://github.com/AdoptOpenJDK/openjdk-build/issues/1212

**Comment by tresf** [comment]
OpenJDK has provided us a much better workaround. https://github.com/AdoptOpenJDK/openjdk-build/issues/1212#issuecomment-521018018

Specifically:


```java
System.setProperty("sun.java2d.print.polling", "false");
```

This seems to be a good compromise.  Lookup is slightly slower (about `90ms`) but it does not suffer the major performance issues introduced with the `AppContext.getAppContext().put(...)` technique.

I'm running it through some performance tests now and it's still performing well (`110ms`) after about 5,000 calls to `lookupPrintServices`.  In contrast, the `AppContext` technique took about `200,000ms` after 100 calls, so this workaround is much more viable.

**Comment by tresf** [comment]
```java
System.setProperty("sun.java2d.print.polling", "false");
```

This does eventually affect performance too, but not nearly as bad.  After about 100,000 calls to `lookupPrintServices`, it adds about 1 second delay to each call.

**Comment by tresf** [comment]
Better workarounds provided 43209f5 (2.1), 
d6cbc7f (2.0).

These fixes will be available in 2.0.11 and 2.1.0-RC8 respectively.

**Comment by tresf** [comment]
Closed via 752df06, 36e971c
Partially related, since it affects search speed #486.

**Comment by tresf** [comment]
QZ Tray 2.0.11 has been released which corrects this problem. https://qz.io/download/
